spec/tbd_tests_spec.rb

```diff
@@ -2794,24 +2647,70 @@
-        s.setAdjacentFoundation(kiva_basement) if surface[:kiva] == :basement
-        s.setAdjacentFoundation(kiva_slab) if surface[:kiva] == :slab
       end
+      expect(found).to be(true)
+    end
+  end
 
-      # Loop through basement wall surfaces and assign foundation object.
-      surfaces.each do |i, surf|
-        next unless found
-        next unless surf.has_key?(:foundation)
-        next unless id == surf[:foundation]
-
-        os_model.getSurfaces.each do |ss|
-          next unless ss.nameString == i
-          next unless ss.outsideBoundaryCondition.downcase == "foundation"
-          ss.setAdjacentFoundation(kiva_basement) if surface[:kiva] == :basement
-          ss.setAdjacentFoundation(kiva_slab) if surface[:kiva] == :slab
-        end
+  it "can generate KIVA exposed perimeters (warehouse)" do
+    translator = OpenStudio::OSVersion::VersionTranslator.new
+    path = OpenStudio::Path.new(File.dirname(__FILE__) + "/files/test_warehouse.osm")
+    os_model = translator.loadModel(path)
+    expect(os_model.empty?).to be(false)
+    os_model = os_model.get
+
+    # Reset all ground-facing floor surfaces as "foundations".
+    os_model.getSurfaces.each do |s|
+      next unless s.outsideBoundaryCondition.downcase == "ground"
+      s.setOutsideBoundaryCondition("Foundation")
+    end
+
+    psi_set = "poor (BC Hydro)"
+    io_path = ""
+    schema_path = File.dirname(__FILE__) + "/../tbd.schema.json"
+    gen_kiva = true
+    io, surfaces = processTBD(os_model, psi_set, io_path, schema_path, gen_kiva)
+    expect(surfaces.size).to eq(23)
+
+    # Validate.
+    surfaces.each do |id, surface|
+      next unless surface.has_key?(:foundation) # only floors
+      next unless surface.has_key?(:kiva)
+      expect(surface[:kiva]).to eq(:slab)
+      expect(surface.has_key?(:exposed)).to be(true)
+      exp = surface[:exposed]
+
+      found = false
+      os_model.getSurfaces.each do |s|
+        next unless s.nameString == id
+        next unless s.outsideBoundaryCondition.downcase == "foundation"
+        found = true
+
+        expect(exp).to be_within(0.01).of(71.62) if id == "Fine Storage"
+        expect(exp).to be_within(0.01).of(35.05) if id == "Office Floor"
+        expect(exp).to be_within(0.01).of(185.92) if id == "Bulk Storage"
       end
+      expect(found).to be(true)
     end
+
+    os_model.save("warehouse_KIVA.osm", true)
+
+    # Now re-open for testing.
+    path = OpenStudio::Path.new(File.dirname(__FILE__) + "/../warehouse_KIVA.osm")
+    os_model2 = translator.loadModel(path)
+    expect(os_model2.empty?).to be(false)
+    os_model2 = os_model2.get
+
+    os_model2.getSurfaces.each do |s|
+      next unless s.isGroundSurface
+      next unless s.nameString == "Fine Storage" ||
+                  s.nameString == "Office Storage" ||
+                  s.nameString == "Bulk Storage"
+      expect(s.outsideBoundaryCondition).to eq("Foundation")
+    end
```

Comment:
This works. But if one runs a simulation, EnergyPlus will crash (missing construction name). Could it be that Kiva'ed surfaces can't rely on construction sets?